forem/forem
# User Settings Step 1: Add user_settings and user_notification_settings tables (rfc#32)

## What type of PR is this? (check all applicable)


## Description
This PR introduce dedicated new tables and corresponding Rails models for storing user-level settings.
This is step one where I introduce the new tables. Hence,  I've simply gone through the User and Profile table and added some of the settings and notification related  attributes to two new tables: 
- users_settings 
- users_notification_settings

The columns are based on [this RFC](https://github.com/forem/rfcs/pull/32/files), however there have been a few amendments (like columns that no longer existed or columns that I think belong in these tables but weren't outlined in the RFC) that I've specifically pointed out in the table below:

**Columns on the `users` table**

|New table| Column name | Usage |
|- | - | - |
| users_settings | `config_font` | "Appearance" on `/settings/customization` |
| users_settings | `config_navbar` | "Appearance" on `/settings/customization` |
| users_settings | `config_theme` | "Appearance" on `/settings/customization` |
| users_settings | `display_announcements` (was not in RFC list) | "Appearance" on `/settings/customization` |
| users_settings | `display_sponsors` (was not in RFC list) | "Appearance" on `/settings/customization` |
| users_settings | `editor_version` | "Writing" on `/settings/customization` |
| users_settings | `experience_level` | "Content" `/settings/customization` |
| users_settings | `feed_admin_publish_permission`| Used in an admin view, but not in a profile view |
| users_settings | `feed_mark_canonical` | "Publishing to DEV Community from RSS" on `/settings/extensions` |
| users_settings | `feed_referential_link` | "Publishing to DEV Community from RSS" on `/settings/extensions` |
| users_settings | `feed_url` | "Publishing to DEV Community from RSS" on `/settings/extensions` |
| users_settings | `inbox_guidelines` | "Connect settings" on `/settings/extensions` |
| users_settings | `inbox_type` | "Connect settings" on `/settings/extensions` |
| users_settings | `language_settings` (this was removed, no longer needed) | "Languages" on `/settings/customization` |
| users_settings | `permit_adjacent_sponsors` | "Content" `/settings/customization` |
| | | | |
| users_notification_settings | `email_badge_notifications` | "Email notifications" on `/settings/notifications` |
| users_notification_settings | `email_comment_notifications` | "Email notifications" on `/settings/notifications` |
| users_notification_settings | `email_community_mod_newsletter` | "Email notifications" on `/settings/notifications` |
| users_notification_settings | `email_connect_messages` | "Email notifications" on `/settings/notifications` |
| users_notification_settings | `email_digest_periodic` | "Email notifications" on `/settings/notifications` |
| users_notification_settings | `email_follower_notifications` | "Email notifications" on `/settings/notifications` |
| users_notification_settings | `email_membership_newsletter` (was not RFC list) | "Email notifications" on `/settings/notifications` |
| users_notification_settings | `email_mention_notifications` | "Email notifications" on `/settings/notifications` |
| users_notification_settings | `email_newsletter` | "Email notifications" on `/settings/notifications` |
| users_notification_settings | `email_tag_mod_newsletter` | "Email notifications" on `/settings/notifications` |
| users_notification_settings | `email_unread_notifications` | "Email notifications" on `/settings/notifications` |
| users_notification_settings | `mobile_comment_notifications` | "Mobile notifications" on `/settings/notifications` |
| users_notification_settings | `mod_roundrobin_notifications` | "Moderator notifications" on `/settings/notifications` |
| users_notification_settings | `reaction_notifications` | "General notifications" on `/settings/notifications` |
| users_notification_settings | `welcome_notifications` | "General notifications" on `/settings/notifications` |

**Profile fields (`attribute_name` column)**

| New table | Profile field attribute name | Usage |
| - | - | - |
| users_settings | `brand_color1` | "Branding" on `/settings/profile` |
| users_settings | `brand_color2` | "Branding" on `/settings/profile` |
| users_settings | `display_email_on_profile` | "Basic" on `/settings/profile` |

TODO:
- [x]  Update the table as there were some additional fields
- [x]  Also need to add the profile fields
- [x] Clean up the code, whitespaces, alphabetize etc.
- [x] Address code climate


## Related Tickets & Documents
https://github.com/forem/rfcs/pull/32/files

## QA Instructions, Screenshots, Recordings
To make sure the models work, you can simply go into the console and do the following:
- `Users::NotificationSetting.create(user_id: User.last.id)`
- `Users::Setting.create(user_id: User.last.id)`

You can take it a step further and add other values to the row. 

### UI accessibility concerns?

_If your PR includes UI changes, please replace this line with details on how
accessibility is impacted and tested. For more info, check out the
[Forem Accessibility Docs](https://docs.forem.com/frontend/accessibility)._

## Added tests?

      have not been included_

## [Forem core team only] How will this change be communicated?

_Will this PR introduce a change that impacts Forem members or creators, the
development process, or any of our internal teams? If so, please note how you
will share this change with the people who need to know about it._

      [Admin Guide](https://forem.gitbook.io/forem-admin-guide/), or
      [Storybook](https://storybook.forem.com/) (for Crayons components)
      or in a [forem.dev](http://forem.dev) post
- [x] This change does not need to be communicated, and this is why not: These new tables are not being used, and even when they are being used, nothing should change for the user

## [optional] Are there any post deployment tasks we need to perform?

## [optional] What gif best describes this PR or how it makes you feel?

![alt_text](https://media.giphy.com/media/kYxjhq6lkT9DoKcLbk/giphy.gif)
